### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized RNN/GRU model with integer arithmetic

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#319]
Public `recipe.train().optimizer(adamw).lr(0.0001).seed(50949421614451).epochs(1).log(all).stop(0.8).fp(8).run()` on `backend:nv0` exceeds the 60-second slow-runtime threshold (stable at 66s). The model (`recipe.model().rnn(8).silu().norm(batch).qi(8).k.gru(8).gelu().qi(3).k.m`) uses quantized `.qi` nodes that lower to `Primitive::Scan` via `lower_scan`, then hang at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA `ptx`/`ptxas` branch (`recipe.rs` ~2852-2864) before any epoch arithmetic begins. The loader (`data/categoric/samples_sqlite.sqlite`) is ignored; the workload size and quantized arithmetic justify sub-second throughput, so the gap is the native compilation stall, not arithmetic or data volume. No existing nv0 issue covers this quantized RNN/GRU scan-combination path.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `39e35844fa022d4b`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized RNN/GRU model with integer arithmetic",
  "body": "Public `recipe.train().optimizer(adamw).lr(0.0001).seed(50949421614451).epochs(1).log(all).stop(0.8).fp(8).run()` on `backend:nv0` exceeds the 60-second slow-runtime threshold (stable at 66s). The model (`recipe.model().rnn(8).silu().norm(batch).qi(8).k.gru(8).gelu().qi(3).k.m`) uses quantized `.qi` nodes that lower to `Primitive::Scan` via `lower_scan`, then hang at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA `ptx`/`ptxas` branch (`recipe.rs` ~2852-2864) before any epoch arithmetic begins. The loader (`data/categoric/samples_sqlite.sqlite`) is ignored; the workload size and quantized arithmetic justify sub-second throughput, so the gap is the native compilation stall, not arithmetic or data volume. No existing nv0 issue covers this quantized RNN/GRU scan-combination path.",
  "rationale": "Inspected `recipe.rs`: `Operation::Rnn` and `Operation::Gru` both lower to `Primitive::Scan` (`lower_scan`, lines 5117-5119, 5442) and reach `compile_native_artifact` on the NVIDIA branch (lines 2852-2864). The packet reports a stable 66-second hang (`elapsed_seconds:66`, `stable:true`, `backend:nv0`) with `.rnn(8).silu().norm(batch).qi(8).k` and `.gru(8).gelu().qi(3).k.m`. Existing nv0 issues cover quantized residual (271), pool/GRU (291), perc/LSTM (302), Bayes/GRU (293), and predictor-only (282), but none represent the earliest failing operation (`compile_native_artifact` before epoch work) through a quantized RNN-plus-GRU scan-node graph. The dataset, seed, activation choices (`silu`/`gelu`), normalization (`norm(batch)`), quantization variants (`.qi(8).k`, `.qi(3).k.m`), loss (`huber`), and `fp(8)` configure the same graph path after or around the failure and do not establish a different earliest cause. Expected runtime: small quantized temporal dataset with 8-wide scan layers justifies sub-second arithmetic; the measured 66s is dominated by the NVIDIA `ptx`/`ptxas` native-artifact hang. Therefore `new`."
}
```

## Deterministic failure packet

kind=performance
id=39e35844fa022d4b
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10049 next:10050 composition:50949421614451
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 66 seconds
measurement=elapsed_seconds:66 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-nv0-10049-335.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/categoric/samples_sqlite.sqlite")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.rnn(8).silu().norm(batch).qi(8).k
		.gru(8).gelu().qi(3).k.m
		.loss(huber);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(50949421614451)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.fp(8)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #428.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #428: Verify current NVIDIA RNN and GRU execution</summary>

Closes #319

This is a verification-only PR. It intentionally changes no tracked files.

## Current public result

Commit: `9098de8711552004bb9eabd84c6d8d9c120d4c58`

Command:

```text
RECIPE_DEBUG=1 ./target/release/recipe --device benji:nv0 /home/nate/.cache/recipe/issues/319/reproduction.rs
```

The exact issue data path, broadcast setting, RNN/GRU model, training settings, and FP8 arithmetic complete successfully:

- The 75,137-row FP8 RNN/GRU artifact begins with `cache=miss`.
- Clang emits NVIDIA LLVM bitcode, and `llc` emits PTX.
- The artifact loads the forward, epoch, and model-load entry points.
- The training epoch completes in 249.111 ms.
- Final Huber loss is finite at 0.1249.
- The process exits 0 after 19.90 seconds.

The reported 66-second NVIDIA native-artifact stall does not reproduce on current `minimal`. No production-code change is supported by this result.


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.

**Comment by nm-z** [comment]
The public `recipe.train().run()` remains active past 60 seconds on nv0 before epoch arithmetic. The earliest failure is native artifact compilation through `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA PTX path. The convolution lowers to a contraction and `.gru(8)` lowers through `lower_scan` to `Primitive::Scan`; this is additional evidence for the existing nv0 GRU compilation stall.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `658a1dd4e5f262de`
- Classifier provider: `copilot`
- Classifier model: `gpt-5.6-luna`
- Classifier effort: `auto`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 319,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized RNN/GRU model with integer arithmetic",
  "body": "The public `recipe.train().run()` remains active past 60 seconds on nv0 before epoch arithmetic. The earliest failure is native artifact compilation through `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA PTX path. The convolution lowers to a contraction and `.gru(8)` lowers through `lower_scan` to `Primitive::Scan`; this is additional evidence for the existing nv0 GRU compilation stall.",
  "rationale": "Issue #319 represents the same earliest public failure and implementation path. Source inspection confirms `.conv(8, 1)` lowers through `lower_conv` and `.gru(8)` through `lower_scan`; `recipe.train().run()` compiles the complete graph before entering the epoch loop. The chronological temporal loader supplies about 12,000 rows and one feature. The graph performs roughly 10 million multiply-accumulate operations, has no tree candidates, depth, or boosting iterations, and moves only a few megabytes per epoch. A defensible nv0 execution range is milliseconds to under one second after compilation, so the measured 60 seconds is materially unjustified and dominated by NVIDIA native artifact compilation. The dataset, fp32 arithmetic, convolution block, normalization, and quantization variants do not establish a different earliest cause."
}
```

## Deterministic failure packet

kind=performance
id=658a1dd4e5f262de
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:23831 next:23832 composition:53529559518229
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-8388-23831/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.conv(8, 1).gelu().norm(batch).qi(4).k.m
		.gru(8).leak().qi(2).k
		.loss(focal);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(53529559518229)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.fp(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0
